exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -491,6 +556,25 @@ private void inferOutputFieldsFromLeftSide() {
       }
     }
 
+    private void inferOutputFieldsFromRightSide() {
```

Comment:
Inferencing happens only once when Drill receives the very first batches from left and right. (Schema-change is not yet supported for Union-All).

Let me summarize the inference in four different situations:
First of all, the field names are always determined by the left side (even when the left side is from empty file, we have the column names. Please see the comment above.)

1. Left: non-empty; Right: non-empty=> types determined by both sides with implicit casting involved
2. Left: empty; Right: non-empty=> type from the right
3. Left: non-empty; Right: empty=> types from the left
4. Left: empty; Right: empty=> types are nullable integer